File: src/main/java/cs/ut/config/nodes/ModelConfiguration.java
package cs.ut.config.nodes;

import cs.ut.config.items.ModelParameter;
import cs.ut.config.items.ModelProperties;
import cs.ut.config.items.Property;
import org.apache.log4j.Logger;

import java.util.*;
import java.util.stream.Collectors;

public class ModelConfiguration {
    private static final Logger log = Logger.getLogger(ModelConfiguration.class);

    private List<String> initalTypes;
    private List<ModelParameter> initialParameters;

    private Map<String, List<ModelParameter>> properties = new LinkedHashMap<>();

    private List<String> basicParameters = new ArrayList<>();

    private ModelConfiguration() {
    }

    public ModelConfiguration(ModelProperties parameters) {
        this.initalTypes = parameters.getTypes();
        this.initialParameters = parameters.getParameters();
        this.basicParameters = parameters.getBasicParams();

        validateAndClassify();
    }

    private void validateAndClassify() {
        assert initalTypes != null : "Expected valid inital types";
        assert initialParameters != null : "Expected valid initial parameters";

        initalTypes.forEach(it -> properties.put(it, new ArrayList<>()));

        initialParameters.forEach(it -> {
            if (initalTypes.contains(it.getType())) {
                properties.get(it.getType()).add(it);
            } else {
                throw new IllegalArgumentException(
                        String.format("Invalid value %s. Allowed parameters are <%s>",
                                it.getType(), initalTypes));
            }
        });

        log.debug("Completed classification for model parameters");
    }

    public Map<String, List<ModelParameter>> getProperties() {
        return new LinkedHashMap<>(properties);
    }

    public List<ModelParameter> getInitialParameters() {
        return this.initialParameters;
    }

    public List<ModelParameter> getBasicParameters() {
        List<ModelParameter> params = new ArrayList<>();

        basicParameters.forEach(param -> {
            Optional<ModelParameter> result = properties
                    .values().stream().flatMap(Collection::stream).filter(it -> param.equals(it.getParameter())).findFirst();
            result.ifPresent(params::add);
        });

        return params;
    }

    public List<Property> getAllProperties() {
        return initialParameters
                .stream()
                .map(ModelParameter::getProperties)
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/cs/ut/config/nodes/ModelConfiguration.kt
package cs.ut.config.nodes

import cs.ut.config.items.ModelParameter
import cs.ut.config.items.Property
import javax.xml.bind.annotation.XmlAccessType
import javax.xml.bind.annotation.XmlAccessorType
import javax.xml.bind.annotation.XmlElement
import javax.xml.bind.annotation.XmlElementWrapper
import javax.xml.bind.annotation.XmlRootElement
import javax.xml.bind.annotation.XmlTransient

@XmlRootElement(name = "modelConfig")
@XmlAccessorType(XmlAccessType.FIELD)
data class ModelConfiguration(
        @field:[
        XmlElementWrapper(name = "types")
        XmlElement(name = "type")]
        var types: MutableList<String>,

        @field:[
        XmlElementWrapper(name = "modelparams")
        XmlElement(name = "modelparam")]
        var parameters: MutableList<ModelParameter>,

        @field:[
        XmlElementWrapper(name = "basicParams")
        XmlElement(name = "param")]
        var basicParams: MutableList<String>) {

    constructor() : this(mutableListOf(), mutableListOf(), mutableListOf())

    @XmlTransient
    lateinit var basicParameters: List<ModelParameter>

    @XmlTransient
    lateinit var properties: Map<String, List<ModelParameter>>

    fun prepareData() {
        properties = parameters.groupBy { it.type }
        assert(types.containsAll(properties.keys), { "Allowed types are $types but received ${properties.keys}" })

        basicParameters = properties.values.flatMap { it }.filter { it.parameter in basicParams }
        assert(basicParameters.isNotEmpty(), { "Basic paramteres are empty but declared values are $basicParams" })
    }

    fun getAllProperties(): List<Property> = parameters.map { it.properties }.flatMap { it }
}